File: crypto/core/rsa_utils.py
"""
RSA Utilities for Hybrid Cryptography System
"""

import os
import base64
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives import serialization, hashes

def generate_rsa_keypair(key_size=2048):
    """Generate RSA public/private key pair"""
    print(f"[INFO] Generating RSA-{key_size} keypair...")
    
    private_key = rsa.generate_private_key(
        public_exponent=65537,
        key_size=key_size
    )
    public_key = private_key.public_key()
    
    print(f"[SUCCESS] RSA-{key_size} keypair generated successfully")
    return private_key, public_key

def save_keys_to_files(private_key, public_key, keys_dir, key_name="rsa_key"):
    """Save RSA keys to PEM files"""
    os.makedirs(keys_dir, exist_ok=True)
    
    # Save private key
    private_pem = private_key.private_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PrivateFormat.PKCS8,
        encryption_algorithm=serialization.NoEncryption()
    )
    
    private_key_path = os.path.join(keys_dir, f"{key_name}_private.pem")
    with open(private_key_path, 'wb') as f:
        f.write(private_pem)
    
    # Save public key
    public_pem = public_key.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo
    )
    
    public_key_path = os.path.join(keys_dir, f"{key_name}_public.pem")
    with open(public_key_path, 'wb') as f:
        f.write(public_pem)
    
    print(f"[SUCCESS] Keys saved to {keys_dir}")
    return private_key_path, public_key_path

def encrypt_aes_key_with_rsa(aes_key, public_key):
    """Encrypt AES key using RSA public key"""
    try:
        encrypted_key = public_key.encrypt(
            aes_key,
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None
            )
        )
        print("[SUCCESS] AES key encrypted with RSA")
        return encrypted_key
    except Exception as e:
        print(f"[ERROR] RSA encryption failed: {e}")
        return None

def decrypt_aes_key_with_rsa(encrypted_aes_key, private_key):
    """Decrypt AES key using RSA private key"""
    try:
        decrypted_key = private_key.decrypt(
            encrypted_aes_key,
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None
            )
        )
        print("[SUCCESS] AES key decrypted with RSA")
        return decrypted_key
    except Exception as e:
        print(f"[ERROR] RSA decryption failed: {e}")
        return None

def load_private_key_from_pem(pem_data):
    """Load RSA private key from PEM string"""
    try:
        if isinstance(pem_data, str):
            pem_data = pem_data.encode('utf-8')
        
        private_key = serialization.load_pem_private_key(
            pem_data,
            password=None
        )
        return private_key
    except Exception as e:
        print(f"[ERROR] Failed to load private key from PEM: {e}")
        return None

def load_public_key_from_pem(pem_data):
    """Load RSA public key from PEM string"""
    try:
        if isinstance(pem_data, str):
            pem_data = pem_data.encode('utf-8')
        
        public_key = serialization.load_pem_public_key(pem_data)
        return public_key
    except Exception as e:
        print(f"[ERROR] Failed to load public key from PEM: {e}")
        return None

def load_private_key_from_file(file_path):
    """Load RSA private key from PEM file"""
    try:
        with open(file_path, 'rb') as f:
            pem_data = f.read()
        return load_private_key_from_pem(pem_data)
    except Exception as e:
        print(f"[ERROR] Failed to load private key from file: {e}")
        return None

def load_public_key_from_file(file_path):
    """Load RSA public key from PEM file"""
    try:
        with open(file_path, 'rb') as f:
            pem_data = f.read()
        return load_public_key_from_pem(pem_data)
    except Exception as e:
        print(f"[ERROR] Failed to load public key from file: {e}")
        return None

def get_key_info(public_key):
    """Get information about RSA key"""
    key_size = public_key.key_size
    return {
        "algorithm": "RSA",
        "key_size_bits": key_size,
        "key_size_bytes": key_size // 8,
        "public_exponent": public_key.public_numbers().e
    }

File: backend/src/main/java/com/sentra/crypto/core/RSAUtils.java
package com.sentra.crypto.core;

import javax.crypto.Cipher;
import javax.crypto.spec.OAEPParameterSpec;
import javax.crypto.spec.PSource;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.KeyFactory;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.PrivateKey;
import java.security.PublicKey;
import java.security.SecureRandom;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.MGF1ParameterSpec;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;
import com.sentra.backend.runtime.Log;

public class RSAUtils {
    public static KeyPair generateRsaKeypair() throws Exception {
        return generateRsaKeypair(2048);
    }

    public static KeyPair generateRsaKeypair(int keySize) throws Exception {
        Log.debug("CRYPTO", "rsa_key", "generating size=" + keySize);
        KeyPairGenerator generator = KeyPairGenerator.getInstance("RSA");
        generator.initialize(keySize);
        KeyPair keyPair = generator.generateKeyPair();
        Log.debug("CRYPTO", "rsa_key", "generated size=" + keySize);
        return keyPair;
    }

    public static Map<String, Path> saveKeysToFiles(PrivateKey privateKey, PublicKey publicKey, Path keysDir, String keyName) throws Exception {
        Files.createDirectories(keysDir);
        String privatePem = toPem("PRIVATE KEY", privateKey.getEncoded());
        String publicPem = toPem("PUBLIC KEY", publicKey.getEncoded());

        Path privatePath = keysDir.resolve(keyName + "_private.pem");
        Path publicPath = keysDir.resolve(keyName + "_public.pem");
        Files.writeString(privatePath, privatePem);
        Files.writeString(publicPath, publicPem);

        Log.debug("CRYPTO", "rsa_key", "saved dir=" + keysDir);

        Map<String, Path> paths = new HashMap<>();
        paths.put("private", privatePath);
        paths.put("public", publicPath);
        return paths;
    }

    public static byte[] encryptAesKeyWithRsa(byte[] aesKey, PublicKey publicKey) {
        try {
            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
            OAEPParameterSpec oaepSpec = new OAEPParameterSpec(
                "SHA-256",
                "MGF1",
                MGF1ParameterSpec.SHA256,
                PSource.PSpecified.DEFAULT
            );
            cipher.init(Cipher.ENCRYPT_MODE, publicKey, oaepSpec);
            byte[] encrypted = cipher.doFinal(aesKey);
            Log.debug("CRYPTO", "rsa_encrypt", "aes_key_wrapped");
            return encrypted;
        } catch (Exception e) {
            Log.error("CRYPTO", "rsa_encrypt", "failed");
            return null;
        }
    }

    public static byte[] encryptAesKeyWithRsaDeterministic(byte[] aesKey, PublicKey publicKey, SecureRandom rng) {
        try {
            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
            OAEPParameterSpec oaepSpec = new OAEPParameterSpec(
                "SHA-256",
                "MGF1",
                MGF1ParameterSpec.SHA256,
                PSource.PSpecified.DEFAULT
            );
            cipher.init(Cipher.ENCRYPT_MODE, publicKey, oaepSpec, rng);
            byte[] encrypted = cipher.doFinal(aesKey);
            Log.debug("CRYPTO", "rsa_encrypt", "aes_key_wrapped");
            return encrypted;
        } catch (Exception e) {
            Log.error("CRYPTO", "rsa_encrypt", "failed");
            return null;
        }
    }

    public static byte[] decryptAesKeyWithRsa(byte[] encryptedAesKey, PrivateKey privateKey) {
        try {
            Cipher cipher = Cipher.getInstance("RSA/ECB/OAEPWithSHA-256AndMGF1Padding");
            OAEPParameterSpec oaepSpec = new OAEPParameterSpec(
                "SHA-256",
                "MGF1",
                MGF1ParameterSpec.SHA256,
                PSource.PSpecified.DEFAULT
            );
            cipher.init(Cipher.DECRYPT_MODE, privateKey, oaepSpec);
            byte[] decrypted = cipher.doFinal(encryptedAesKey);
            Log.debug("CRYPTO", "rsa_decrypt", "aes_key_unwrapped");
            return decrypted;
        } catch (Exception e) {
            Log.error("CRYPTO", "rsa_decrypt", "failed");
            return null;
        }
    }

    public static PrivateKey loadPrivateKeyFromPem(String pem) throws Exception {
        byte[] der = parsePem(pem);
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePrivate(new PKCS8EncodedKeySpec(der));
    }

    public static PublicKey loadPublicKeyFromPem(String pem) throws Exception {
        byte[] der = parsePem(pem);
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePublic(new X509EncodedKeySpec(der));
    }

    public static PrivateKey loadPrivateKeyFromFile(Path filePath) throws Exception {
        String pem = Files.readString(filePath);
        return loadPrivateKeyFromPem(pem);
    }

    public static PublicKey loadPublicKeyFromFile(Path filePath) throws Exception {
        String pem = Files.readString(filePath);
        return loadPublicKeyFromPem(pem);
    }

    public static Map<String, Object> getKeyInfo(PublicKey publicKey) {
        Map<String, Object> info = new HashMap<>();
        if (publicKey instanceof RSAPublicKey) {
            RSAPublicKey rsa = (RSAPublicKey) publicKey;
            info.put("algorithm", "RSA");
            info.put("key_size_bits", rsa.getModulus().bitLength());
            info.put("key_size_bytes", rsa.getModulus().bitLength() / 8);
            info.put("public_exponent", rsa.getPublicExponent());
        }
        return info;
    }

    private static String toPem(String type, byte[] derBytes) {
        String base64 = Base64.getEncoder().encodeToString(derBytes);
        StringBuilder sb = new StringBuilder();
        sb.append("-----BEGIN ").append(type).append("-----\n");
        for (int i = 0; i < base64.length(); i += 64) {
            int end = Math.min(i + 64, base64.length());
            sb.append(base64, i, end).append("\n");
        }
        sb.append("-----END ").append(type).append("-----\n");
        return sb.toString();
    }

    private static byte[] parsePem(String pem) {
        String[] lines = pem.replace("\r", "").split("\n");
        StringBuilder sb = new StringBuilder();
        for (String line : lines) {
            if (line.startsWith("-----")) {
                continue;
            }
            sb.append(line.trim());
        }
        return Base64.getDecoder().decode(sb.toString());
    }
}
